peps/pep-0740.rst

```diff
@@ -205,110 +181,340 @@ changes to it:
+      """
+
+  @dataclass
+  class Provenance:
+      version: Literal[1]
+      """
+      The provenance object's version, which is always 1.
+      """
+
+      attestations: list[tuple[Publisher, list[Attestation]]]
+      """
+      One or more attestation objects, indexed by their publisher.
+      """
+
+* ``version`` is ``1``. Like attestation objects, provenance objects are
+  versioned, and this PEP only defines version ``1``.
+* ``attestations`` is a **required** JSON array, containing pairs
+  of publishers and a JSON array of one or more attestation objects.
+
+  Each publisher is a Trusted Publishing identity, as described
+  by the ``Publisher`` model above. As noted in the ``Publisher`` model,
+  each publisher object is specific to its Trusted Publisher but must include
+  at minimum:
+
+  * A ``kind`` key, which **MUST** be a JSON string that uniquely identifies the
     kind of Trusted Publisher.
   * A ``claims`` key, which **MUST** be a JSON object containing any context-specific
     claims retained by the index during Trusted Publisher authentication.
 
-  All other keys in the ``publisher`` object are publisher-specific. A full
-  illustrative example of a ``publisher`` object is provided in :ref:`appendix-2`.
-* ``attestations`` is a **required** JSON object, containing one or
-  more attestation objects as identified by their keys. This object is
-  a superset of ``attestations`` object supplied by the uploader through the
-  ``attestations`` field, as described in :ref:`upload-endpoint`.
+  All other keys in the publisher object are publisher-specific. A full
+  illustrative example of a publisher object is provided in :ref:`appendix-1`.
 
-  Because ``attestations`` is a superset of the file's original uploaded attestations,
-  the index **MAY** chose to embed additional attestations of its own.
+  Each array of attestation objects is a superset of the ``attestations``
+  array supplied by the uploaded through the ``attestations`` field at upload
+  time, as described in :ref:`upload-endpoint` and
+  :ref:`changes-to-provenance-objects`.
 
-Simple Index
-^^^^^^^^^^^^
+.. _changes-to-provenance-objects:
 
-* When an uploaded file has one or more attestations, the index **MAY** include a
-  ``data-provenance`` attribute on its file link, with a value of either
-  ``true`` or ``false``.
-* When ``data-provenance`` is ``true``, the index **MUST** serve a
-  :ref:`provenance object <provenance-object>` at the same URL, but with
-  ``.provenance`` appended to it. For example, if ``HolyGrail-1.0.tar.gz``
-  exists and has associated attestations, those attestations would be located
-  within the provenance object hosted at ``HolyGrail-1.0.tar.gz.provenance``.
+Changes to provenance objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-JSON-based Simple API
-^^^^^^^^^^^^^^^^^^^^^
+Provenance objects are *not* immutable, and may change over time. Reasons
+for changes to the provenance object include but are not limited to:
```

Comment:
The examples mentioned below of the provenance objects changing are only about _adding_ new things (attestations, publishers). Does that mean that provenance objects are mutable, but append-only? Or can attestations be _removed_ too?

Reply:
As specified, they're append-only. I can imagine scenarios in which the index _might_ want to remove attestations however, so I didn't want to formally guarantee that the index never does so.